packages/useful-types/src/types.ts

```diff
@@ -4,3 +4,10 @@ export type ThenType<T> = T extends Promise<infer U> ? U : T;
 export type FirstArgument<T> = T extends ((arg: infer U) => any) ? U : never;
 export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;
 export type Props<T> = T extends ComponentType<infer P> ? P : never;
+export type DeepPartial<T> = {
+  [P in keyof T]?: T[P] extends Array<infer U>
+    ? Array<DeepPartial<U>>
+    : T[P] extends ReadonlyArray<infer U>
+      ? ReadonlyArray<DeepPartial<U>>
```

Comment:
I couldn't get the `ReadonlyArray` to infer properly, have you tried testing that out?

Reply:
i'll try and build a type set that covers this issue

Reply:
Resolved via conversations on slack. Merging as is.